src/ui/features/referral-program/ReferralProgramService.background.ts

```diff
@@ -0,0 +1,157 @@
+import { INTERNAL_ORIGIN } from 'src/background/constants';
+import { readSavedReferrerData, saveReferrerData } from './shared/storage';
+
+interface Options {
+  getWallet: () => Wallet;
+}
+
+class ReferralProgramService {
+  private options: Options | null = null;
+
+  initialize(options: Options) {
+    this.options = options;
+    emitter.on('walletCreated', this.applyReferralCodeToWallet.bind(this));
+  }
+
+  async signReferralMessage({
+    address,
+    referralCode,
+  }: {
+    address: string;
+    referralCode: string;
+  }) {
+    invariant(this.options, "Options aren't expected to become null");
+
+    const walletFacade = this.options.getWallet();
+
+    const normalizedAddress = normalizeAddress(address);
+    const message = `${normalizedAddress} -> ${referralCode}`;
+
+    const signature = await walletFacade.signMessage({
+      params: {
+        signerAddress: address,
+        message,
+        messageContextParams: {
+          initiator: INTERNAL_ORIGIN,
+          clientScope: null,
+        },
+      },
+      context: { origin: INTERNAL_ORIGIN },
+    });
+
+    return signature;
+  }
+
+  async signAndSendReferWalletMessage({
+    address,
+    referralCode,
+  }: {
+    address: string;
+    referralCode: string;
+  }) {
+    const signature = await this.signReferralMessage({
+      address,
+      referralCode,
+    });
+    return ZerionAPI.referWallet({
+      address,
+      referralCode,
+      signature,
+    });
+  }
+
+  async applyReferralCodeToAllWallets({
+    referralCode,
+  }: {
+    referralCode: string;
+  }) {
+    invariant(this.options, "Options aren't expected to become null");
+    const walletFacade = this.options.getWallet();
+
+    const checkReferralResponse = await ZerionAPI.checkReferral({
+      referralCode,
+    });
+    const checkedReferrer = checkReferralResponse.data;
+    const walletGroups = await walletFacade.uiGetWalletGroups({
+      context: INTERNAL_SYMBOL_CONTEXT,
+    });
+
+    const signerWalletGroups = walletGroups?.filter((group) =>
+      isSignerContainer(group.walletContainer)
+    );
+    const ownedBareWallets =
+      signerWalletGroups?.flatMap(
+        (group) =>
+          group.walletContainer.wallets.filter(isBareWallet) as BareWallet[]
+      ) || [];
+    const ownedAddresses = ownedBareWallets.map((wallet) => wallet.address);
+
+    const walletsMetaResponse =
+      ownedAddresses.length > 0
+        ? await ZerionAPI.getWalletsMeta({ identifiers: ownedAddresses })
+        : null;
+    const walletsMeta = walletsMetaResponse?.data || [];
+
+    const eligibleAddresses = walletsMeta
+      .filter(
+        (meta) =>
+          meta.membership.referrer === null &&
+          meta.membership.referralCode !== referralCode
+      )
+      .map((meta) => meta.address);
+
+    await Promise.allSettled(
+      eligibleAddresses.map((address) =>
+        this.signAndSendReferWalletMessage({ address, referralCode })
+      )
+    );
```

Comment:
Do we handle failed requests anywhere?

Reply:
Not yet. We agreed I’d send `client_error` messages for failed requests. I’ll update this part in a few minutes